test/distributed/elastic/agent/server/test/local_elastic_agent_test.py

```diff
@@ -963,6 +971,59 @@ def test_double_agent_fault_tolerance_etcd_v2(self):
             backend="etcd-v2", test_to_run=self.double_agent_fault_tolerance
         )
 
+    def no_exit_barrier_on_failure(self):
+        """
+        start ``nnodes`` agents, kill and restart odd ones, validate fault-tolerance works
+        """
+        nnodes = 2
+        wait = 20
+        node_conf = Conf(
+            entrypoint=_bipolar_sleep_function, args=(wait,), local_world_size=2
+        )
+        agent_results = mp.Queue()
+        agent_args = {
+            "conf": node_conf,
+            "agent_results": agent_results,
+            "min_nodes": nnodes,
+            "max_nodes": nnodes,
+            "max_restarts": 0,
+            "exit_barrier_timeout": 300,
+        }
+
+        procs = []
+        for _ in range(nnodes):
+            p = mp.Process(
+                target=self.run_agent,
+                kwargs=agent_args,
+            )
+            procs.append(p)
+            p.start()
+
+        # wait for all processes to finish should not take exit barrier time
+        exit_interval_between_agents = 0
+        for i in range(nnodes):
+            p = procs[i]
+            p.join()
+            self.assertNotEqual(0, p.exitcode)
+            exit_interval_between_agents = (
+                time.monotonic() - exit_interval_between_agents
+            )
+
+        # Validate that the processes finish close to each other.
+        # Using a slightly higher timeout than 2 * monitor_interval (0.01) to make it less flaky
+        monitor_interval = 0.1
```

Comment:
you'd have to pass this `monitor_interval` as the agent config when you run the agent, or obtain the config that you ran the agent with and use the monitor_interval in the config object.

The way you have it here assumes that the agent was kicked off with `monitor_interval=0.1` which isn't quite true b/c the [`get_worker_spec()`](https://github.com/pytorch/pytorch/blob/main/test/distributed/elastic/agent/server/test/local_elastic_agent_test.py#L274) function defaults to `0.01` and `run_agent()` doesn't expose the `monitor_interval` flag (so you'll have to plumb it in)

Reply:
So, i plumbed the values. But the test remained flaky. The test needs more than 2 * monitor_interval for some values of monitor_interval, i stress tested with monitor_interval = 0.1 , 0.5 seconds and found that it did fail with 0.1  (1 failure out of 20 runs) , with 0.5 all my tests succeeded ( 20 stress tests), but i could not find logic that guaranteed shutdown within 2 * monitor_interval 

https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/agent/server/api.py#L701. Anyways, updated the diff with this suggestion and monitor_interval = 0.5 s

Reply:
cc @kiukchung  PTAL